qa/vagrant/src/main/java/org/elasticsearch/packaging/util/Packages.java

```diff
@@ -90,13 +93,27 @@ public static Installation install(Distribution distribution) {
     }
 
     public static Installation install(Distribution distribution, String version) {
+        final Result result = runInstallCommand(distribution, version);
+        if (result.exitCode != 0) {
+            throw new RuntimeException("Installing distribution " + distribution + " version " + version + " failed: " + result);
+        }
+
+        return Installation.ofPackage(distribution.packaging);
+    }
+
+    public static Result runInstallCommand(Distribution distribution) {
+        return runInstallCommand(distribution, getCurrentVersion());
+    }
+
+    public static Result runInstallCommand(Distribution distribution, String version) {
         final Shell sh = new Shell();
         final Path distributionFile = getDistributionFile(distribution, version);
 
-        Platforms.onRPM(() -> sh.run("rpm -i " + distributionFile));
-        Platforms.onDPKG(() -> sh.run("dpkg -i " + distributionFile));
-
-        return Installation.ofPackage(distribution.packaging);
+        if (Platforms.isRPM()) {
```

Comment:
nit: the more I see of this, the more I think that it's worth extending the `onDPKG` type of API to be able to produce values so reading trough the code is more consistent.

Reply:
Thanks. The version of that I tried looked like this

```java
Result result = Platforms.onRpm("...")
if (result == null) {
  Platforms.onDpkg("...")
}
return result;
```

because the `onFoo` methods still have to return something when they don't run their action. Which doesn't seem much more readable to me. Doing it with an `Optional` is a little better. Is there a better one that I'm missing?

Reply:
```
   Result result = Platforms.packagingChoice()
        .onRpm("...")
        .onDpkg()
        .get()
```
The `packagingChoice()` is optional, only if you don't want to keep the Platforms class growing.
The `onRpm()`   etc type of methods will an instance, say `ResultTracker<Result>` and depending of the choice to be made will or will not add a result to it. `get()` will just return the result or throw an exception. We could also have a `ResultTracker<Void>` and a `decide()` method instead of `get()` that doesn't return anything. Maybe `PackagingDecisionMaker` is a better name than `ResultTracker`. Since this is for testing, we could have all the result trackers/decision makers register into some static context in `Platforms` perhaps, and add a check after each test to make sure that all of them are "decided" to make sure that one doesn't `Platforms.onmRpm().onDpkg()` without forgetting to call `decide()`.